Trim managedFields in controller-manager

linxiulei opened the pull request:
#### What type of PR is this?

/kind feature


#### What this PR does / why we need it:

Improve memory efficiency for `kube-controller-manager`

#### Which issue(s) this PR fixes:

Fixes #118454

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
Improve memory usage of kube-controller-manager by dropping the `.metadata.managedFields` field that kube-controller-manager doesn't require.
```

linxiulei committed:
Add 'WithTransform' SharedInformerOption

Also add support for metadata informer factory to use SharedInformerOption

Signed-off-by: Eric Lin <exlin@google.com>

linxiulei committed:
Trim managedFields in controller-manager

Signed-off-by: Eric Lin <exlin@google.com>

linxiulei committed:
Run hack/update-codegen.sh informers subprojects

Signed-off-by: Eric Lin <exlin@google.com>

linxiulei commented:
/cc @wojtek-t @apelisse @alexzielenski

linxiulei commented on staging/src/k8s.io/controller-manager/pkg/informerfactory/informer_factory.go:
```diff
@@ -42,6 +43,18 @@ func (i *informerFactory) ForResource(resource schema.GroupVersionResource) (inf
 }
 
 func (i *informerFactory) Start(stopCh <-chan struct{}) {
+	trimFunc := func(obj interface{}) (interface{}, error) {
```
@apelisse this code removes ManagedFields, PTAL.

apelisse replied:
Ah, yes, that's what I missed, thanks :-)

alexzielenski commented on staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:
```diff
@@ -257,6 +257,17 @@ func (f *sharedInformerFactory) InformerFor(obj {{.runtimeObject|raw}}, newFunc
 }
 
+// ExistingInformers returns all existing shared informers.
+func (f *sharedInformerFactory) ExistingInformers() map[{{.schemaGroupVersionResource|raw}}]GenericInformer {
```
Rather than add `ExistingInformers` to the factory, I personally would have added a new `SharedInformerOption` that allows the user to provide a closure like `func(informers.GenericInformer)` to run against the informer when the informer is created by the factory in `InformerFor`

linxiulei replied:
Great suggestion. I added 'WithTransform' SharedInformerOption to make this option more precise.

czybjtu commented:
should we modify every informer according on your way which we don't use managedFileld?

alexzielenski commented:
/triage accepted

linxiulei commented:
> should we modify every informer according on your way which we don't use managedFileld?

Yes, that'd be my plan

linxiulei commented:
Cleaned up and ready to review. I'd appreciate a "ok-to-test", thanks in advance.

apelisse commented:
/ok-to-test

apelisse commented on staging/src/k8s.io/client-go/metadata/metadatainformer/informer.go:
```diff
@@ -31,6 +32,33 @@ import (
+// SharedInformerOption defines the functional option type for metadataSharedInformerFactory.
+type SharedInformerOption func(*metadataSharedInformerFactory) *metadataSharedInformerFactory
+
+// WithTweakListOptions sets a custom filter on all listers of the configured metadataSharedInformerFactory.
+func WithTweakListOptions(tweakListOptions TweakListOptionsFunc) SharedInformerOption {
+	return func(factory *metadataSharedInformerFactory) *metadataSharedInformerFactory {
+		factory.tweakListOptions = tweakListOptions
+		return factory
+	}
+}
+
+// WithNamespace limits the metadataSharedInformerFactory to the specified namespace.
+func WithNamespace(namespace string) SharedInformerOption {
+	return func(factory *metadataSharedInformerFactory) *metadataSharedInformerFactory {
+		factory.namespace = namespace
+		return factory
+	}
+}
```
How is this related?

linxiulei replied:
No, this is not related. But I'd like to make metadataSharedInformerFactory have identical SharedInformerOption(s) with SharedInformerFactory for consistency.

apelisse commented:
Can we make them in separate PRs then? I'd rather do one thing at a time. Happy to lgtm/approve otherwise.

linxiulei commented:
> Can we make them in separate PRs then? I'd rather do one thing at a time. Happy to lgtm/approve otherwise.

Sure, I removed unnecessary changes and rebased. Thanks!

linxiulei commented:
/retest

apelisse commented:
thanks
/lgtm
/approve

wojtek-t commented:
/lgtm
/approve

linxiulei commented:
/retest

alculquicondor commented:
/hold

For the implication of stripping the managedFields when
- KCM controllers use Update
- Out-of-tree controllers use SSA.

alculquicondor commented:
FYI,
I'm supporting this change for kube-scheduler https://github.com/kubernetes/kubernetes/pull/119556

However, the change in kcm needs approval from SIG Architecture, IIUC.

alculquicondor commented:
Ah, @wojtek-t's approval should suffice
/hold cancel

aojea commented:
This seems better to wait for next release in 1.29, seems like a change risky for this part of the release cycle

/cc @liggitt

alculquicondor commented:
Agree. As it stands, this is currently blocked on the code freeze, as intended.

liggitt commented:
A data race on resync was just reported in use of WithTransform in https://github.com/kubernetes/client-go/issues/1348 that may affect this use

wojtek-t commented:
heh...
the problem is that for Resync operation we take object that is already existing in the store (DeltaFifo in this case).
And there is chance of touching (reading) that object at the same time.

We effectively can't call Transform for SYNC events here:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L456

liggitt commented:
> We effectively can't call Transform for SYNC events here:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L456

Or delete events where the object is DeletedTombstoneFinalStateUnknown, and maybe also for Replace (the docs there are unclear)

nirvana274 commented:
> > Or delete events where the object is DeletedFinalStateUnknown, and maybe also for Replace (the docs there are unclear - "In cases involving Replace(), such an object can come through multiple times.")
> 
> Good point on the DeletedFinalStateUnknown.
> 
> Regarding Replace() - this is probably true in theory, but in practice we only replace with a net new result of List call, so in practice it's not a problem [although we should clarify the contract there or sth]
